File: JHUSpinWidthPaper_2015/LHEAnalyzer/macros/trimPythia.cc
//std includes
#include <iostream>
#include <utility>
#include <vector>
//root includes
#include "TFile.h"
#include "TLorentzVector.h"
#include "TString.h"
#include "TTree.h"
//CMSSW includes
#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
#include "DataFormats/JetReco/interface/GenJet.h"
#include "FWCore/Utilities/interface/TypeWithDict.h"
#include "SimDataFormats/GeneratorProducts/interface/GenEventInfoProduct.h"

using namespace std;

vector<pair<int, int>> findDuplicates(const vector<double>* fourvector, vector<int> id, vector<int> status);

void trimPythia(TString cinput, TString outdir="./", int fileLevel=1){
  TString coutput = outdir;
  coutput.Append("pythiaTemp.root");

  TFile* ftemp = new TFile(coutput, "recreate");
  TTree* tmpTree = new TTree("tmpTree", "");

  vector<double> geneventinfoweights;

  vector<double> reco_GenJet_FV[4];
  vector<int> reco_GenJet_id;
  vector<int> reco_GenJet_status;

  vector<double> reco_GenParticle_FV[4];
  vector<int> reco_GenParticle_id;
  vector<int> reco_GenParticle_status;

  tmpTree->Branch("genWeights", &geneventinfoweights);

  tmpTree->Branch("reco_GenJet_X", reco_GenJet_FV);
  tmpTree->Branch("reco_GenJet_Y", reco_GenJet_FV+1);
  tmpTree->Branch("reco_GenJet_Z", reco_GenJet_FV+2);
  tmpTree->Branch("reco_GenJet_E", reco_GenJet_FV+3);
  tmpTree->Branch("reco_GenJet_id", &reco_GenJet_id);
  tmpTree->Branch("reco_GenJet_status", &reco_GenJet_status);

  tmpTree->Branch("reco_GenParticle_X", reco_GenParticle_FV);
  tmpTree->Branch("reco_GenParticle_Y", reco_GenParticle_FV+1);
  tmpTree->Branch("reco_GenParticle_Z", reco_GenParticle_FV+2);
  tmpTree->Branch("reco_GenParticle_E", reco_GenParticle_FV+3);
  tmpTree->Branch("reco_GenParticle_id", &reco_GenParticle_id);
  tmpTree->Branch("reco_GenParticle_status", &reco_GenParticle_status);



  TFile f(cinput, "read");
  if (!f.IsZombie()){
    TTree* events = (TTree*)f.Get("Events");
    events->SetBranchStatus("*", 0);

    edm::Wrapper< vector<reco::GenJet> >* jetWrapper;
    edm::Wrapper< vector<reco::GenParticle> >* genparticleWrapper;
    edm::Wrapper< GenEventInfoProduct >* geneventinfoWrapper;

    TString suffix;
    if (fileLevel == 1) suffix = "SIM";
    else if (fileLevel == 2) suffix = "GEN";
    else {
      cout << "trimPythia should not be called with fileLevel=" << fileLevel << endl;
      assert(0);
    }
    events->SetBranchStatus("recoGenJets_ak5GenJets__"+suffix+"*", 1);
    events->SetBranchAddress("recoGenJets_ak5GenJets__"+suffix+".", &jetWrapper);
    events->SetBranchStatus("recoGenParticles_genParticles__"+suffix+"*", 1);
    events->SetBranchAddress("recoGenParticles_genParticles__"+suffix+".", &genparticleWrapper);
    events->SetBranchStatus("GenEventInfoProduct_generator__"+suffix+"*", 1);
    events->SetBranchAddress("GenEventInfoProduct_generator__"+suffix+".", &geneventinfoWrapper);

    for (int ev=0; ev<events->GetEntries(); ev++){
      events->GetEntry(ev);

      geneventinfoweights.clear();

      for (int v=0; v<4; v++){
        reco_GenJet_FV[v].clear();
        reco_GenParticle_FV[v].clear();
      }
      reco_GenJet_id.clear();
      reco_GenParticle_id.clear();
      reco_GenJet_status.clear();
      reco_GenParticle_status.clear();

      const gen::PdfInfo* genpdf = geneventinfoWrapper->product()->pdf();
      const vector<double> genweights = geneventinfoWrapper->product()->weights();
      for (unsigned int g=0; g<genweights.size(); g++) geneventinfoweights.push_back(genweights.at(g));

      const vector<reco::GenJet>* reco_GenJets = jetWrapper->product();
      const vector<reco::GenParticle>* reco_GenParticles = genparticleWrapper->product();

      for (unsigned int p=0; p<reco_GenParticles->size(); p++){
        const reco::GenParticle& part = reco_GenParticles->at(p);
        if (
          ((part.pdgId()==25 || part.pdgId()==32) && part.status()==22) // Generated Higgs
          ||
          ((std::abs(part.pdgId())>=11 && std::abs(part.pdgId())<=16) && (part.status()==23 || part.status()==1)) // Generated leptons
          ||
          ((std::abs(part.pdgId())<=6 || std::abs(part.pdgId())==21) && (part.status()==21 || part.status()==23 || part.status()==22)) // Generated partons
          ){
          reco_GenParticle_FV[0].push_back(part.px());
          reco_GenParticle_FV[1].push_back(part.py());
          reco_GenParticle_FV[2].push_back(part.pz());
          reco_GenParticle_FV[3].push_back(part.energy());
          reco_GenParticle_id.push_back(part.pdgId());
          reco_GenParticle_status.push_back(part.status());
        }
      }

      vector<pair<int, int>> reco_GenParticle_duplicates = findDuplicates(reco_GenParticle_FV, reco_GenParticle_id, reco_GenParticle_status);

      // Add status==1 or status==23 (with precedence for status==1) leptons to genJet collection
      for (unsigned int p=0; p<reco_GenParticle_id.size(); p++){
        if ((std::abs(reco_GenParticle_id.at(p))>=11 && std::abs(reco_GenParticle_id.at(p))<=16) && std::abs(reco_GenParticle_id.at(p)) % 2 == 1){
          bool match=false;
          for (unsigned int dd=0; dd<reco_GenParticle_duplicates.size(); dd++){
            pair<int, int> duplicate = reco_GenParticle_duplicates.at(dd);
            int iOriginal = duplicate.first; // Status==23 particle
            if (iOriginal==int(p)){ match=true; break; }
          }
          if (match) continue;
          for(int fv=0;fv<4;fv++) reco_GenJet_FV[fv].push_back(reco_GenParticle_FV[fv].at(p));
          reco_GenJet_id.push_back(reco_GenParticle_id.at(p));
          reco_GenJet_status.push_back(reco_GenParticle_status.at(p));
        }
      }

      vector<int> removalArray;
      for (unsigned int dd=0; dd<reco_GenParticle_duplicates.size(); dd++){
        pair<int, int> duplicate = reco_GenParticle_duplicates.at(dd);
        int iTransfer = duplicate.second; // Status==1 particle
        bool inserted=false;
        for (unsigned int it = 0; it<removalArray.size(); it++){
          int iIndex = removalArray.at(it);
          if (iTransfer > iIndex){
            removalArray.insert(removalArray.begin()+it, iTransfer);
            inserted=true;
            break;
          }
        }
        if (!inserted) removalArray.push_back(iTransfer);
      }
      // Remove status==1 duplicates from genParticles
      for (unsigned int dd=0; dd<removalArray.size(); dd++){
        int iTransfer = removalArray.at(dd);
        for (int fv=0; fv<4; fv++){
          reco_GenParticle_FV[fv].erase(reco_GenParticle_FV[fv].begin()+iTransfer);
        }
        reco_GenParticle_id.erase(reco_GenParticle_id.begin()+iTransfer);
        reco_GenParticle_status.erase(reco_GenParticle_status.begin()+iTransfer);
      }

      for (unsigned int p=0; p<reco_GenJets->size(); p++){
        const reco::GenJet& jet = reco_GenJets->at(p);
        reco_GenJet_FV[0].push_back(jet.px());
        reco_GenJet_FV[1].push_back(jet.py());
        reco_GenJet_FV[2].push_back(jet.pz());
        reco_GenJet_FV[3].push_back(jet.energy());
        reco_GenJet_id.push_back(jet.pdgId());
        reco_GenJet_status.push_back(jet.status());
      }

      tmpTree->Fill();
    }
    f.Close();
  }
  else if (f.IsOpen()) f.Close();
  ftemp->WriteTObject(tmpTree);
  delete tmpTree;
  ftemp->Close();
}

vector<pair<int, int>> findDuplicates(const vector<double>* fourvector, vector<int> id, vector<int> status){
  vector<pair<int, int>> duplicates;

  for (unsigned int xx=0; xx<id.size(); xx++){
    TLorentzVector p_tm(fourvector[0].at(xx), fourvector[1].at(xx), fourvector[2].at(xx), fourvector[3].at(xx));
    int id_tm = id.at(xx);
    int st_tm = status.at(xx);
    if (st_tm==23 || st_tm==1){
      for (unsigned int yy=xx+1; yy<id.size(); yy++){
        TLorentzVector p_tbm(fourvector[0].at(yy), fourvector[1].at(yy), fourvector[2].at(yy), fourvector[3].at(yy));
        int id_tbm = id.at(yy);
        int st_tbm = status.at(yy);

        if (id_tbm==id_tm && st_tbm!=st_tm && (st_tbm==1 || st_tbm==23)){
          double dot_tm = p_tbm.Dot(p_tm);
          double diff_sqmass = dot_tm - p_tm.M2();
          diff_sqmass = fabs((double)diff_sqmass);
          if (diff_sqmass<0.005*fabs(p_tm.M2())){
            int iFirst = (st_tm==23 ? xx : yy); // Should be order-independent
            int iSecond = (st_tbm==1 ? yy : xx);
            pair<int, int> dupinst(iFirst, iSecond);
            duplicates.push_back(dupinst);
          }
        }
      }
    }
  }

  return duplicates;
}

File: JHUSpinWidthPaper_2015/LHEAnalyzer/macros/trimPythia.py
import loadLib
import os
import ROOT
import sys

def trimPythia(cinput, outdir, pythiaStep, jetAlgorithm):
  pythiaStep = int(pythiaStep)
  coutput = os.path.join(outdir, "pythiaTemp.root")

  ftemp = ROOT.TFile(coutput, "recreate")
  tmpTree = ROOT.TTree("tmpTree", "")

  geneventinfoweights = ROOT.vector("double")()

  reco_GenJet_FV = [ROOT.vector("double")() for i in range(4)]
  reco_GenJet_id = ROOT.vector("int")()
  reco_GenJet_status = ROOT.vector("int")()

  reco_GenParticle_FV = [ROOT.vector("double")() for i in range(4)]
  reco_GenParticle_id = ROOT.vector("int")()
  reco_GenParticle_status = ROOT.vector("int")()

  tmpTree.Branch("genWeights", "vector<double>", ROOT.AddressOf(geneventinfoweights))

  tmpTree.Branch("reco_GenJet_X", "vector<double>", ROOT.AddressOf(reco_GenJet_FV[0]))
  tmpTree.Branch("reco_GenJet_Y", "vector<double>", ROOT.AddressOf(reco_GenJet_FV[1]))
  tmpTree.Branch("reco_GenJet_Z", "vector<double>", ROOT.AddressOf(reco_GenJet_FV[2]))
  tmpTree.Branch("reco_GenJet_E", "vector<double>", ROOT.AddressOf(reco_GenJet_FV[3]))
  tmpTree.Branch("reco_GenJet_id", "vector<int>", ROOT.AddressOf(reco_GenJet_id))
  tmpTree.Branch("reco_GenJet_status", "vector<int>", ROOT.AddressOf(reco_GenJet_status))

  tmpTree.Branch("reco_GenParticle_X", "vector<double>", ROOT.AddressOf(reco_GenParticle_FV[0]))
  tmpTree.Branch("reco_GenParticle_Y", "vector<double>", ROOT.AddressOf(reco_GenParticle_FV[1]))
  tmpTree.Branch("reco_GenParticle_Z", "vector<double>", ROOT.AddressOf(reco_GenParticle_FV[2]))
  tmpTree.Branch("reco_GenParticle_E", "vector<double>", ROOT.AddressOf(reco_GenParticle_FV[3]))
  tmpTree.Branch("reco_GenParticle_id", "vector<int>", ROOT.AddressOf(reco_GenParticle_id))
  tmpTree.Branch("reco_GenParticle_status", "vector<int>", ROOT.AddressOf(reco_GenParticle_status))



  f = ROOT.TFile(cinput, "read")
  if not f.IsZombie():
    events = f.Get("Events")
    events.SetBranchStatus("*", 0)

    if pythiaStep == 1: suffix = "SIM"
    elif pythiaStep == 0: suffix = "GEN"
    else:
      print "trimPythia should not be called with pythiaStep=%i" % pythiaStep
      assert False
    events.SetBranchStatus("recoGenJets_"+jetAlgorithm+"GenJets__"+suffix+"*", 1)
    events.SetBranchStatus("recoGenParticles_genParticles__"+suffix+"*", 1)
    events.SetBranchStatus("GenEventInfoProduct_generator__"+suffix+"*", 1)

    for ev in events:
      geneventinfoweights.clear()

      for v in range(4):
        reco_GenJet_FV[v].clear()
        reco_GenParticle_FV[v].clear()
      reco_GenJet_id.clear()
      reco_GenParticle_id.clear()
      reco_GenJet_status.clear()
      reco_GenParticle_status.clear()

      geneventinfoWrapper = getattr(ev, "GenEventInfoProduct_generator__"+suffix+".")
      genpdf = geneventinfoWrapper.product().pdf()
      genweights = geneventinfoWrapper.product().weights()
      for genweight in genweights: geneventinfoweights.push_back(genweight)

      jetWrapper = getattr(ev, "recoGenJets_"+jetAlgorithm+"GenJets__"+suffix+".")
      genparticleWrapper = getattr(ev, "recoGenParticles_genParticles__"+suffix+".")
      reco_GenJets = jetWrapper.product()
      reco_GenParticles = genparticleWrapper.product()

      for part in reco_GenParticles:
        if (
          ((part.pdgId()==25 or part.pdgId()==32) and part.status()==22) # Generated Higgs
          or
          ((abs(part.pdgId())>=11 and abs(part.pdgId())<=16) and (part.status()==23 or part.status()==1)) # Generated leptons
          or
          ((abs(part.pdgId())<=6 or abs(part.pdgId())==21) and (part.status()==21 or part.status()==23 or part.status()==22)) # Generated partons
          ):
          reco_GenParticle_FV[0].push_back(part.px())
          reco_GenParticle_FV[1].push_back(part.py())
          reco_GenParticle_FV[2].push_back(part.pz())
          reco_GenParticle_FV[3].push_back(part.energy())
          reco_GenParticle_id.push_back(part.pdgId())
          reco_GenParticle_status.push_back(part.status())

      reco_GenParticle_duplicates = findDuplicates(reco_GenParticle_FV, reco_GenParticle_id, reco_GenParticle_status)

      # Add status==1 or status==23 (with precedence for status==1) leptons to genJet collection
      for p, id in enumerate(reco_GenParticle_id):
        if (abs(id)>=11 and abs(id) <= 16 and abs(id) % 2 == 1):
          match = False
          for iOriginal, iDuplicate in reco_GenParticle_duplicates:
            if iOriginal==id:
              match = True
              break
          if match: continue
          for fv in range(4): reco_GenJet_FV[fv].push_back(reco_GenParticle_FV[fv].at(p))
          reco_GenJet_id.push_back(id)
          reco_GenJet_status.push_back(reco_GenParticle_status.at(p))

      removalArray = []
      for iFirst, iTransfer in reco_GenParticle_duplicates:
        # iTransfer is the status==1 particle
        inserted = False
        for it, iIndex in enumerate(removalArray):
          if iTransfer > iIndex:
            removalArray.insert(it, iTransfer)
            inserted = True
            break
        if not inserted: removalArray.append(iTransfer)
      # Remove status==1 duplicates from genParticles
      for iTransfer in removalArray:
        for fv in reco_GenParticle_FV:
          fv.erase(fv.begin()+iTransfer)
        reco_GenParticle_id.erase(reco_GenParticle_id.begin()+iTransfer)
        reco_GenParticle_status.erase(reco_GenParticle_status.begin()+iTransfer)

      for jet in reco_GenJets:
        reco_GenJet_FV[0].push_back(jet.px())
        reco_GenJet_FV[1].push_back(jet.py())
        reco_GenJet_FV[2].push_back(jet.pz())
        reco_GenJet_FV[3].push_back(jet.energy())
        reco_GenJet_id.push_back(jet.pdgId())
        reco_GenJet_status.push_back(jet.status())

      tmpTree.Fill()
    f.Close()

  elif f.IsOpen(): f.Close()
  ftemp.WriteTObject(tmpTree)
  del tmpTree
  ftemp.Close()

def findDuplicates(fourvectors, ids, statuses):
  #fourvectors = [pxs, pys, pzs, Es], convert to list of 4 momenta
  tlorentzvectors = [ROOT.TLorentzVector(*pxpypzE) for pxpypzE in zip(*fourvectors)]
  assert len(tlorentzvectors) == len(ids) == len(statuses)
  duplicates = []

  for xx, (p_tm, id_tm, st_tm) in enumerate(zip(tlorentzvectors, ids, statuses)):
    if st_tm==23 or st_tm==1:
      for yy, (p_tbm, id_tbm, st_tbm) in enumerate(zip(tlorentzvectors[xx+1:], ids[xx+1:], statuses[xx+1:]), start=xx+1):
        if (id_tbm==id_tm and st_tbm!=st_tm and (st_tbm==1 or st_tbm==23)):
          dot_tm = p_tbm.Dot(p_tm)
          diff_sqmass = abs(dot_tm - p_tm.M2())
          if (diff_sqmass<0.005*abs(p_tm.M2())):
            iFirst = (xx if st_tm==23 else yy) #Should be order-independent
            iSecond = (yy if st_tbm==1 else xx)
            duplicates.append((iFirst, iSecond))

  return duplicates

if __name__ == "__main__":
    trimPythia(*sys.argv[1:5])
